examples/plot_gp_function.py

```diff
@@ -0,0 +1,237 @@
+    utility_ei = utility_function_ei.utility(x, None, optimizer2._gp, 0)
+    
+    utility_function_poi = UtilityFunction(kind="poi", kappa=5, xi=0)
+    utility_poi = utility_function_poi.utility(x, None, optimizer3._gp, 0)
+    
+    #utility_function_kg = UtilityFunction(kind="kg", kappa=5, xi=0)
+    #utility_kg = utility_function_kg.utility(x, optimizer, optimizer._gp, 0)
+    
+    # UCB
+    acq.plot(x, utility_ucb, label='UCB', color='purple')
+    acq.plot(x[np.argmax(utility_ucb)], np.max(utility_ucb), '*', markersize=15, markerfacecolor='purple', markeredgecolor='k', markeredgewidth=1)
+    
+    # EI
+    acq.plot(x, utility_ei, label='EI', color='green')
+    acq.plot(x[np.argmax(utility_ei)], np.max(utility_ei), '*', markersize=15, markerfacecolor='green', markeredgecolor='k', markeredgewidth=1)
+    
+    # Poi
+    acq.plot(x, utility_poi, label='POI', color='orange')
+    acq.plot(x[np.argmax(utility_poi)], np.max(utility_poi), '*', markersize=15, markerfacecolor='orange', markeredgecolor='k', markeredgewidth=1)
+    
+    # KG
+    #acq.plot(x, utility_kg, label='KG', color='orange')
+    #acq.plot(x[np.argmax(utility_kg)], np.max(utility_kg), '*', markersize=15, markerfacecolor='orange', 
+             #markeredgecolor='k', markeredgewidth=1)
+    
+    
+    acq.set_xlim((min(x), max(x)))
+    acq.set_ylim((min(min(utility_ucb), min(utility_ei), min(utility_poi)) - 0.5, max(max(utility_ucb), max(utility_ei), max(utility_poi)) + 0.5))
+    acq.set_ylabel('Utility', fontdict={'size':20})
+    acq.set_xlabel('x', fontdict={'size':20})
+    
+    #plt.plot(x, y, color='grey', label='Objective function', linewidth=4, linestyle=':' )
+    #plt.axvline(x=x[np.argmax(y)], linestyle=':')
+    
+    axis.legend(loc=2, bbox_to_anchor=(1.01, 1), borderaxespad=0.)
+    acq.legend(loc=2, bbox_to_anchor=(1.01, 1), borderaxespad=0.)
+    
+    plt.show()
+    
+    
+    
+def plot_convergence(optimizer1,optimizer2,optimizer3, x, target):
+    x = x.reshape(-1,1)
+    y = target(x)
+    
+    #def suggest(utility, target, optimizer):
+        #next_point = optimizer.suggest(utility)
+        #target = target(**next_point)
+        #optimizer.register(params=next_point, target=target)
+    
+        #return target, next_point
+        #print(optimizer.max)
+    it=20
+    tar1=np.zeros(it)
+    tar2=np.zeros(it)
+    tar3=np.zeros(it)
+    point1=np.zeros(it)
+    point2=np.zeros(it)
+    point3=np.zeros(it)
+    opt=None
+    #if kind=='kg':
+        #opt=optimizer
+        
+    for i in range(it):
+        utility_function1 = UtilityFunction(kind='ucb', kappa=5, xi=0)
+        utility_function2 = UtilityFunction(kind='ei', kappa=5, xi=0)
+        utility_function3 = UtilityFunction(kind='poi', kappa=5, xi=0)
+        utility1 = utility_function1.utility(x, opt, optimizer1._gp, 0)
+        utility2 = utility_function2.utility(x, opt, optimizer2._gp, 0)
+        utility3 = utility_function3.utility(x, opt, optimizer3._gp, 0)
```

Comment:
Qui state passando `ymax=0` e non capisco perché. La documentazione in `util.py` ma anche la definizione di POI ed EI vi dicono che `ymax` è l'incumbent, cioè il miglior valore trovato finora. Che sia questa la causa dell'errore?

Reply:
Abbiamo aggiornato ymax ma i grafici non cambiano.
Abbiamo però notato che il problema potrebbe essere legato al fatto che Poi è costante (=1) quindi il massimo calcolato per questa acquisition è casuale, quindi non corrisponde a quello ritornato dalla funzione suggest. 
In generale, **solo** nelle iterazioni in cui le acquisition functions non hanno un massimo unico riscontriamo questo problema.

Reply:
Ho realizzato ora che effettivamente dichiarate le variabili `utility1:3` ma non le utilizzate mai... È voluto? In ogni caso, non capisco allora quale sia il problema del grafico

Reply:
Anche la POI costante è qualcosa che dovreste indagare. Magari avete messo dei parametri troppo laschi alla funzione di acquisizione. Provate a cambiarli e a vedere se trovate valori più adatti. A proposito di parametri, potreste aggiungere a `plot_convergence()` un argomento: un dizionario di dizionari che contenga i valori dei parametri da assegnare alle varie acqusizioni, cosicché possano essere controllati "da fuori". Questo vi faciliterebbe anche il testing di vari parametri. Per esempio:
```python
params = {'ucb': {'kappa': 5, 'xi': 0}, 'ei': {...}, 'poi': {...}}
plot_convergence(optimizer1,optimizer2,optimizer3, x, target, params)
```

Reply:
Quelle variabili non vengono utilizzate perchè ci siamo accorti dopo "suggest" non vuole la utility ma direttamente la utility_function (quindi errore nostro, le abbiamo lasciate ma non servono). 
Per quanto riguarda il grafico, dai plot delle acquisition functions abbiamo dei "suggested x" che non corrispondo ai "suggested x" plottati nel grafico finale delle convergenze... e questo si può notare bene per Poi, in cui per esempio in un'iterazione il valore suggerito diventa negativo (grafico "after 5 steps") nel grafico delle acquisitions ma nel grafico delle convegenze rimangono sempre tutti positivi (notebook plot_acq_func_examples).

Reply:
Se effettivamente la POI è costante, il punto potrebbe essere scelto casualmente. La divergenza può essere quindi spiegata dai diversi insiemi di numeri pseudo-casuali generati. Non so quanto controllo abbiamo su questa cosa, ma sicuramente si risolve a monte rendendo la POI non costante. Anche perché in tal caso viene meno l'intero punto dell'utilizzo di una funzione di acquisizione. Ribadisco quindi l'importanza di sperimentare con diversi valori dei parametri: `kappa` (eventualmente anche `kappa_decay`) e `xi`. Sono pur sempre informazioni interessanti e rilevanti per il progetto, e quindi per la presentazione/report!

Reply:
La funzione non era costante, ma con variazioni troppo piccole per essere viste nello stesso grafico delle altre. Abbiamo risolto mettendo nuovi parametri per Poi e normalizzando le utlity facendo in modo che variassero tutte nello stesso range (0,1).
Può vedere il file aggiornato con i grafici "acq_func_examples.ipynb" nella cartella examples.